pkg/scheduler/factory_test.go

```diff
@@ -273,145 +272,66 @@ func TestCreateFromConfigWithUnspecifiedPredicatesOrPriorities(t *testing.T) {
-		return client.CoreV1().Pods(namespace).Get(context.TODO(), name, metav1.GetOptions{})
-	}))
-	// podLister := corelisters.NewPodLister(podIndexer)
-	stopCh := make(chan struct{})
-	defer close(stopCh)
-
-	timestamp := time.Now()
-	queue := internalqueue.NewPriorityQueue(nil, internalqueue.WithClock(clock.NewFakeClock(timestamp)))
-	schedulerCache := internalcache.New(30*time.Second, stopCh)
-	errFunc := MakeDefaultErrorFunc(client, podLister, queue, schedulerCache)
-
 	tests := []struct {
 		name                       string
-		initFn                     func()
-		deferFn                    func()
 		injectErr                  error
 		podUpdatedDuringScheduling bool // pod is updated during a scheduling cycle
 		podDeletedDuringScheduling bool // pod is deleted during a scheduling cycle
-		expectNode                 *v1.Node
 		expect                     *v1.Pod
-		testBackoff                bool
 	}{
 		{
-			name:       "node is deleted during a scheduling cycle",
-			initFn:     func() { schedulerCache.AddNode(nodeFoo) },
-			injectErr:  apierrors.NewNotFound(apicore.Resource("node"), nodeFoo.Name),
-			expectNode: nodeFoo,
-		},
-		{
-			name: "pod is updated during a scheduling cycle",
-			initFn: func() {
-				queue.Add(testPod)
-				queue.Pop()
-			},
-			deferFn:                    func() { queue.Delete(testPod) },
+			name:                       "pod is updated during a scheduling cycle",
+			injectErr:                  nil,
 			podUpdatedDuringScheduling: true,
 			expect:                     testPodUpdated,
 		},
 		{
-			name: "pod is not updated during a scheduling cycle",
-			initFn: func() {
-				queue.Add(testPod)
-				queue.Pop()
-			},
-			deferFn: func() { queue.Delete(testPod) },
-			expect:  testPod,
-		},
-		{
-			name: "pod is not updated during a scheduling cycle",
-			initFn: func() {
-				queue.Add(testPod)
-				queue.Pop()
-			},
-			deferFn:     func() { queue.Delete(testPod) },
-			expect:      testPod,
-			testBackoff: true,
+			name:      "pod is not updated during a scheduling cycle",
+			injectErr: nil,
+			expect:    testPod,
 		},
 		{
-			name: "pod is deleted during a scheduling cycle",
-			initFn: func() {
-				queue.Add(testPod)
-				queue.Pop()
-			},
-			deferFn:                    func() { queue.Delete(testPod) },
+			name:                       "pod is deleted during a scheduling cycle",
+			injectErr:                  nil,
 			podDeletedDuringScheduling: true,
 			expect:                     nil,
 		},
 	}
 
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			if tt.initFn != nil {
-				tt.initFn()
-			}
-			if tt.deferFn != nil {
-				defer tt.deferFn()
-			}
+			stopCh := make(chan struct{})
+			defer close(stopCh)
+
+			client := fake.NewSimpleClientset(&v1.PodList{Items: []v1.Pod{*testPod}})
+			informerFactory := informers.NewSharedInformerFactory(client, 0)
+			podInformer := informerFactory.Core().V1().Pods()
+			// Need to add/update/delete testPod to the store.
```

Comment:
what about the following?

```
	cc.InformerFactory.Start(ctx.Done())
	cc.InformerFactory.WaitForCacheSync(ctx.Done())
```

Reply:
I tried but didn't help. Say `podLister.Pods(pod.Namespace).Get(pod.Name)` won't return the expected cache pod.

Reply:
Did you do it after creating the podInformer?

Reply:
I tried both (https://gist.github.com/Huang-Wei/8fdabdcafd4d8614fe748c66986e995a) and even used the clientset to Create() the testPod. In all cases, `podLister.Pods(pod.Namespace).Get(pod.Name)` returns nil 😞

Reply:
Aha, I found the trick here:

- `informerFactory` doesn't instantiate any informer when it's created (informers.NewSharedInformerFactory)
- `podInformer`, create by `informerFactory.Core().V1().Pods()` is nothing but a "hollow" interface - the concrete podInfomer and its indexer are not created yet
- when the real podInformer (and its indexer) is created? You're hearing right - by explicitly call `podInfomer.Informer()` or `podInfomer.Indexer()` - which will instantiate everything a podInformer needs. Otherwise, the podInformer will be only ready until call ErrFunc (by calling the informer's Lister() function), but by that time, informerFactory has already started for a while, so the informer doesn't really work.
- given the above fact, I really don't want the test to rely on such a tricky triggering point of podInformer.

Reply:
This is standard. I remember running into this in the past when writing my own tests.

Since this is what we do in production code anyways, we can just do the same.